### PR TITLE
use static variable to do cache instead of thread local in thread frequent switching case

### PR DESCRIPTION
**Pull request by LeoZhao-BJ**
This case is special, NativeExecutor Predictor.run is called in a thread frequent switching case, for mkldnn, TransferScope cache creation is unavoidable. While current cache mechanism is using thread local variable to store cache, if execution thread is changed every iteration, it will always create new memory which makes big memory leak.

The solution is to use a global static variable to handle this case by explicitly indicating from user. Here I reuse SetMkldnnThreadid.

fix https://github.com/PaddlePaddle/Paddle/pull/18372#issuecomment-506929844

### CODE REVIEW

File: paddle/fluid/framework/transfer_scope_cache.cc [hunk 1]
```diff
@@ -17,14 +17,31 @@
 namespace paddle {
 namespace framework {
 
+static std::unordered_map<size_t, Scope*>* static_transfer_data_cache = nullptr;
+static std::unordered_set<Scope*>* static_transfer_scope_cache = nullptr;
+
 std::unordered_map<size_t, Scope*>& global_transfer_data_cache() {
-  thread_local auto* x = new std::unordered_map<size_t, Scope*>;
-  return *x;
+  // if get_cur_thread_id() == -1, means not use thread local method to do cache
```
Comment (luotao1) [inline paddle/fluid/framework/transfer_scope_cache.cc]: Add more comments `platform::get_cur_thread_id() == -1` is only used in MKLDNN.
OR: 
```
#ifdef PADDLE_WITH_CUDA
    thread_local auto* x = new std::unordered_map<size_t, Scope*>;
    return *x;
#else
#ifdef PADDLE_WITH_MKLDNN
   if (platform::get_cur_thread_id() == -1) {
      if (!static_transfer_data_cache)
         static_transfer_data_cache = new std::unordered_map<size_t, Scope*>;
   }
#endif
#endif
```
?

Comment (LeoZhao-BJ) [reply paddle/fluid/framework/transfer_scope_cache.cc]: good idea, to be updated.

Comment (LeoZhao-BJ) [reply paddle/fluid/framework/transfer_scope_cache.cc]: it is not special mkldnn issue, but this solution only works with mkldnn api.

**Comment by LeoZhao-BJ** [comment]
Using static variable solution and API is to make it simple, otherwise if we use single globe structure, it makes more complicated, it need thread sync etc, and efficiency will drop.

**Comment by luotao1** [comment]
@LeoZhao-Intel, Discussed with @Superjomn :
1) Do you have some comments for "-1", or do you have some unit-test to ensure this "-1"?
2) Do you need add lock for `static_transfer_scope_cache`? When multi-instance use "-1" mode, it will fail on this.

**Comment by LeoZhao-BJ** [comment]
> @LeoZhao-Intel, Discussed with @Superjomn :
> 
> 1. Do you have some comments for "-1", or do you have some unit-test to ensure this "-1"?

-1 is just specific for mkldnn single instance but with thread frequent switching, so far we don't have unit-test.

> 2. Do you need add lock for `static_transfer_scope_cache`? When multi-instance use "-1" mode, it will fail on this.

In mkldnn single instance but with thread frequent switching, it only supports one instance, yes, it will fail if with multi-instances. We do have plan to refine this solution in the future.

**Comment by jczaja** [comment]
@luotao1 , @Superjomn Why GPU execution needs to store TransferScope in TLS, why each thread has to keep its own copy of scopes and data caches eg. what is wrong with having only one global data and scope cache?

**Comment by luotao1** [comment]
> Why GPU execution needs to store TransferScope in TLS? why each thread has to keep its own copy of scopes and data caches eg. what is wrong with having only one global data and scope cache?

@jczaja CUDADeviceContext is thread_local, i.e. one GPU-card have one CUDADeviceContext.

File: paddle/fluid/framework/transfer_scope_cache.cc [hunk 2]
```diff
@@ -18,42 +18,62 @@ namespace paddle {
...
-    return *static_transfer_data_cache;
-  } else {
-#endif
-    thread_local auto* x = new std::unordered_map<size_t, Scope*>;
-    return *x;
-#ifdef PADDLE_WITH_MKLDNN
+  size_t tid = static_cast<size_t>(platform::get_cur_thread_id());
+
+  // if tid != 0, means there is specific mkldnn tid setting from user.
+  if (tid != 0) {
```
Comment (luotao1) [inline paddle/fluid/framework/transfer_scope_cache.cc]: How about use CONST value to describe "0" or "-1" mode, such as `constexpr char kEnableCacheRuntimeContext` etc. You can put these CONST value somewhere.

File: paddle/fluid/framework/transfer_scope_cache.cc [hunk 3]
```diff
@@ -17,12 +17,61 @@
...
+using transfer_scope_cache_map = std::unordered_set<Scope*>;
+static std::unordered_map<size_t, transfer_data_cache_map*>
+    static_transfer_data_caches;
+static std::unordered_map<size_t, transfer_scope_cache_map*>
+    static_transfer_scope_caches;
+#endif
+
 std::unordered_map<size_t, Scope*>& global_transfer_data_cache() {
+#ifdef PADDLE_WITH_MKLDNN
+  size_t tid = static_cast<size_t>(platform::get_cur_thread_id());
```
Comment (luotao1) [inline paddle/fluid/framework/transfer_scope_cache.cc]: get_cur_thread_id->get_cur_mkldnn_session_id

File: paddle/fluid/framework/transfer_scope_cache.cc [hunk 4]
```diff
@@ -17,12 +17,61 @@
...
+
+    auto map_it = static_transfer_scope_caches.find(tid);
+    if (map_it == static_transfer_scope_caches.end()) {
+      auto* x = new transfer_scope_cache_map;
+      static_transfer_scope_caches[tid] = x;
+      return *x;
+    } else {
+      return *static_transfer_scope_caches[tid];
+    }
+  }
```
Comment (luotao1) [inline paddle/fluid/framework/transfer_scope_cache.cc:73]: I will help you add a unit-test here.

**Comment by LeoZhao-BJ** [comment]
I guess, it is similar with mkldnn cache, to support multi-instance, thread_local is used to simply the implementation.

File: paddle/fluid/framework/transfer_scope_cache.cc (hunk 3 above)
Comment (LeoZhao-BJ) [reply paddle/fluid/framework/transfer_scope_cache.cc]: new PR #18453 is submitted

File: paddle/fluid/framework/transfer_scope_cache.cc (hunk 2 above)
Comment (LeoZhao-BJ) [reply paddle/fluid/framework/transfer_scope_cache.cc]: change in PR #18453 .

Comment (LeoZhao-BJ) [reply paddle/fluid/framework/transfer_scope_cache.cc]: solved

File: paddle/fluid/framework/transfer_scope_cache.cc (hunk 3 above)
Comment (LeoZhao-BJ) [reply paddle/fluid/framework/transfer_scope_cache.cc]: solved

File: paddle/fluid/inference/tests/api/analyzer_bert_tester.cc [hunk 5]
```diff
@@ -251,6 +251,7 @@ TEST(Analyzer_bert, transfer_scope_cache) {
     threads.emplace_back([&, i]() {
       std::getline(fin, line);
       ParseLine(line, &input);
+      if (is_static) platform::set_cur_mkldnn_session_id(1);
```
Comment (luotao1) [inline paddle/fluid/inference/tests/api/analyzer_bert_tester.cc:257]: Please add a TODO here, we will hide directly use of `platform::set_cur_mkldnn_session_id(1)` later.

File: paddle/fluid/inference/tests/api/analyzer_bert_tester.cc [hunk 6]
```diff
@@ -264,9 +265,21 @@ TEST(Analyzer_bert, transfer_scope_cache) {
...
   // paddle::framework::global_transfer_data_cache() are thread_local,
   // their pointer should be different among different thread id.
-  PADDLE_ENFORCE(global_transfer_scope_cache.size(), threads_num);
-  PADDLE_ENFORCE(global_transfer_data_cache.size(), threads_num);
+  if (is_static) {
+    PADDLE_ENFORCE(global_transfer_scope_cache.size(), 1);
+    PADDLE_ENFORCE(global_transfer_data_cache.size(), 1);
+  } else {
+    PADDLE_ENFORCE(global_transfer_scope_cache.size(), threads_num);
+    PADDLE_ENFORCE(global_transfer_data_cache.size(), threads_num);
```
Comment (luotao1) [inline paddle/fluid/inference/tests/api/analyzer_bert_tester.cc:281]: line 265-267 is the comment for line 272 and 273. please add comments for  line269-270

File: paddle/fluid/inference/tests/api/analyzer_bert_tester.cc (hunk 5 above)
Comment (LeoZhao-BJ) [reply paddle/fluid/inference/tests/api/analyzer_bert_tester.cc:257]: done

File: paddle/fluid/inference/tests/api/analyzer_bert_tester.cc (hunk 6 above)
Comment (LeoZhao-BJ) [reply paddle/fluid/inference/tests/api/analyzer_bert_tester.cc:281]: done

**Review by luotao1 — DISMISSED** [review]
LGTM

**Comment by luotao1** [comment]
```
[14:29:15]	/home/teamcity/work/e84e6e698a3f913d/paddle/fluid/inference/tests/api/analyzer_bert_tester.cc:256:32: error: no member named 'set_cur_mkldnn_session_id' in namespace 'paddle::platform'
[14:29:15]	      if (is_static) platform::set_cur_mkldnn_session_id(1);
[14:29:15]	                     ~~~~~~~~~~^
```
In mac_ci

File: paddle/fluid/inference/tests/api/analyzer_bert_tester.cc [hunk 7]
```diff
@@ -261,12 +264,26 @@ TEST(Analyzer_bert, transfer_scope_cache) {
...
     std::vector<PaddleTensor>().swap(input);
   }
-  // Since paddle::framework::global_transfer_scope_cache() and
-  // paddle::framework::global_transfer_data_cache() are thread_local,
-  // their pointer should be different among different thread id.
-  PADDLE_ENFORCE(global_transfer_scope_cache.size(), threads_num);
-  PADDLE_ENFORCE(global_transfer_data_cache.size(), threads_num);
+  if (is_static) {
+    // Use static method to do transfer_scope_cache() instead of thread_local
+    // so paddle::framework::global_transfer_data_cache() should be same
```
Comment (luotao1) [inline paddle/fluid/inference/tests/api/analyzer_bert_tester.cc]: should be same->should always be 1

Comment (LeoZhao-BJ) [reply paddle/fluid/inference/tests/api/analyzer_bert_tester.cc]: done

**Review by luotao1 — APPROVED** [review]
LGTM

**Comment by luotao1** [comment]
@LeoZhao-Intel Could you revert this PR since #18578 merge and there is no random fail CI on MKLDNN?

**Comment by LeoZhao-BJ** [comment]
> @LeoZhao-Intel Could you revert this PR since #18578 merge and there is no random fail CI on MKLDNN?

ok, will revert it soon.

**Comment by LeoZhao-BJ** [comment]
see new PR #18879 for revert,

**Pull request merged**